### ISSUE TITLE
Error 500 after updated webui2 'Contacts' object has no attribute 'password'

### ISSUE THREAD

**Issue by jbaptl** [shinken-monitoring/mod-webui#543]
Hello

After updated webui2, it became imposible to login on the web interface. 
Moreover, the login page is unusual.

So, i updated all the composants (now i use python2.7) and reinstaled last version of shinken and of his modules, but i steel have the same error.
When I log, i have an AttributeError : `AttributeError("'Contacts' object has no attribute 'password'",)`
The admin contact is well created on contacts dir with password

See attached file for the visual aspect of the login page.
![shinken_login](https://cloud.githubusercontent.com/assets/7994745/18705057/ebbaaff0-7fec-11e6-8382-ab73c892ab57.png)

Here my traceback message after login: 

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/bottle.py", line 862, in _handle
    return route.call(**args)
  File "/usr/local/lib/python2.7/site-packages/bottle.py", line 1732, in wrapper
    rv = callback(*a, **ka)
  File "/var/lib/shinken/modules/webui2/module.py", line 521, in lock_version
    return f(**args)
  File "/var/lib/shinken/modules/webui2/plugins/login/login.py", line 117, in user_auth
    is_authenticated = app.check_authentication(login, password)
  File "/var/lib/shinken/modules/webui2/module.py", line 811, in check_authentication
    user = self.auth_module.check_auth(username, password)
  File "/var/lib/shinken/modules/webui2/submodules/auth.py", line 37, in check_auth
    if self.check_cfg_password_auth(user, password):
  File "/var/lib/shinken/modules/webui2/submodules/auth.py", line 63, in check_cfg_password_auth
    p = c.password
AttributeError: 'Contacts' object has no attribute 'password'
```

And the logs errors :

```
[1474446582] INFO: [broker-master] [WebUI]  user '' is signing in ...
[1474446582] INFO: [broker-master] [WebUI] Checking authentication for user: 
[1474446582] INFO: [broker-master] [WebUI] Requesting authentication for user: 
[1474446582] INFO: [broker-master] [WebUI] Authenticating user ''
[1474446582] INFO: [broker-master] [WebUI] Authenticating user '' with auth-cfg-password
[1474446582] INFO: [broker-master] [Cfg Password UI] User  try to init
[1474446582] WARNING: [broker-master] [WebUI] The mod auth-cfg-password raised an exception: 'Contacts' object has no attribute 'password', I'm tagging it to restart later
[1474446582] WARNING: [broker-master] [WebUI] Exception type: <type 'exceptions.AttributeError'>
[1474446582] WARNING: [broker-master] Back trace of this kill: Traceback (most recent call last):
  File "/var/lib/shinken/modules/webui2/submodules/auth.py", line 27, in check_auth
    if mod.check_auth(user, password):
  File "/var/lib/shinken/modules/auth-cfg-password/module.py", line 75, in check_auth
    p = c.password
AttributeError: 'Contacts' object has no attribute 'password'

[1474446582] INFO: [broker-master] [WebUI] Embedded authentication for ''
[1474446582] INFO: [broker-master] [WebUI-auth-cfg-password] Authenticating user ''
[1474446583] INFO: [broker-master] Trying to init module: auth-cfg-password
[1474446583] INFO: [broker-master] [Cfg Password UI] Trying to initialize the CFG/Password auth
```

The user is not defined although it was informed.
I suspecte a bottle problem, because i have lot of GET info message when i restart shinken :

```
[...]
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/forms/change_var/<name:path>' {}>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/widget/notification_toggle' {}>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/static/notifications/:path#.+#' '/var/lib/shinken/modules/webui2/plugins/notifications'>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/hosts-tags' {}>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/static/tags/:path#.+#' '/var/lib/shinken/modules/webui2/plugins/tags'>
[..]
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/static/:path#.+#' <webui2.module.Webui_broker object at 0x7f92c65239d0>>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/favicon.ico' <function give_favicon at 0x7f92b409f488>>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/opensearch.xml' <webui2.module.Webui_broker object at 0x7f92c65239d0>>
[1474447682] INFO: [broker-master] [WebUI] route: <GET '/modal/:path#.+#' <function give_modal at 0x7f92b409f758>>
[1474447682] INFO: [broker-master] [WebUI] starting Web UI server on 0.0.0.0:7767 ...
[...]
```

I'm using shinken 2.4.3 with webui2 2.4.2c on CentOS6.6

> python2.7
> bottle-0.12.8
> pymongo-3.3.0
> requests-2.11.1
> arrow-0.8.0
> passlib-1.6.5
> CherryPy-8.1.0
> pycurl-7.43.0
> pbr-1.10.0

For the tests, i only use webui2 and auth_cfg_password modules with default configuration.

I really thank you for your help. It could very help me because i'm totaly lost :-)
Also, thank you for the job!

**Comment by jbaptl** [comment]
![login_shinken](https://cloud.githubusercontent.com/assets/7994745/18704996/92e08c42-7fec-11e6-930d-a969fa39a09d.png)

**Comment by mohierf** [comment]
You should not use the `auth_cfg_password` with the WebUI2 ... this module is no more necessary because it is now embedded into the WebUI module itself ... remove this module from you webui2.cfg file and I think that everything will go fine :wink: 

Did you see this installation guide: https://github.com/shinken-monitoring/mod-webui/wiki/Installing-WebUI-authentication-modules?

**Comment by jbaptl** [comment]
I thank you for you quick feedback.
So, i tried to login after removed the `auth_cfg_password` and also after add the `auth_active_directory` module (configuration of this module worked well before the update) , but i still have the error, and the user is still not get in both case. 
For sur i restarted shinken after each modifications. :-(

**Comment by mohierf** [comment]
Ok. Looking more accurately in your logs, I see that the `user` is an empty field !

```
[1474446582] INFO: [broker-master] [WebUI]  user '' is signing in ...
[1474446582] INFO: [broker-master] [WebUI] Checking authentication for user: 
[1474446582] INFO: [broker-master] [WebUI] Requesting authentication for user: 
[1474446582] INFO: [broker-master] [WebUI] Authenticating user ''
[1474446582] INFO: [broker-master] [WebUI] Authenticating user '' with auth-cfg-password
```

You are trying to log-in with an empty username field from the login form ... I invite you to clear your Web browser cache ... and to confirm that the username name is present in the login form. it is certainly the source of this problem.

**Comment by jbaptl** [comment]
I clear the web browser cache and cookies on differents browser (firefox and chrome) on different computers, but i still have the same error. 
The logs still indicate empty field, but i confirm that I set correctly the user and the password present on contact file.
I 'm sure, it is something very stupid but i spended lot of time on it and i'm out of idea.
What do you thinkd about the `route : <GET` messages on broker logs? It seems unusual.

Thanks

**Comment by mohierf** [comment]
The `route:` logs are simply here to indicate the configured Bottle routes and messages is a valid one ...

I suppose that you do not have any front-end server before the WebUI port ?

Should you have a look in the browser's development console what is really posted to the webui when you submit the login form ?

**Comment by jbaptl** [comment]
No device can block the route. 
To validated this point, i tried with a browser on localhost and i still have the error.

What is really posted seems to be good. 
Here a screenshot :

![formulaire_shinken](https://cloud.githubusercontent.com/assets/7994745/18712016/bc96608c-800c-11e6-999a-73acf8965262.PNG)

Is it what you espected for?

**Comment by mohierf** [comment]
Ok. Everything seems ok on this part. 

I do not understand why you do not have an 'admin' user information trying to authenticate in the webui log ...

I just tested on my environment (Debian) and got this in the log:

```
[1474467781] INFO: [broker-master] [WebUI]  user 'admin' is signing in ...
[1474467781] INFO: [broker-master] [WebUI] Checking authentication for user: admin
[1474467781] INFO: [broker-master] [WebUI] Requesting authentication for user: admin
[1474467781] INFO: [broker-master] [WebUI] Authenticating user 'admin'
[1474467781] INFO: [broker-master] [WebUI] Internal htpasswd authentication
[1474467781] INFO: [broker-master] [WebUI] Internal alignak backend authentication
[1474467781] INFO: [broker-master] [WebUI] Internal contact authentication
[1474467781] INFO: [broker-master] [WebUI-auth-cfg-password] Authenticating user 'admin'
[1474467781] INFO: [broker-master] [WebUI-auth-cfg-password] Authenticated
[1474467781] INFO: [broker-master] [WebUI] user authenticated thanks to contact
```

Should you copy the same part of your log?

**Comment by jbaptl** [comment]
Sur, It is the same you refered upper :

```
[1474467590] INFO: [broker-master] [WebUI]  user '' is signing in ...
[1474467590] INFO: [broker-master] [WebUI] Checking authentication for user: 
[1474467590] INFO: [broker-master] [WebUI] Requesting authentication for user: 
[1474467590] INFO: [broker-master] [WebUI] Authenticating user ''
[1474467590] INFO: [broker-master] [WebUI] Embedded authentication for ''
[1474467590] INFO: [broker-master] [WebUI-auth-cfg-password] Authenticating user ''
```

**Comment by mohierf** [comment]
You log is slightly different than mine ... I do not find **Embedded authentication** in the most recent source code :/

 And, you post on /user/login with login and password fields. This is managed by a function that:

```
def user_auth():
    login = app.request.forms.get('login', '')
    password = app.request.forms.get('password', '')
    logger.info("[WebUI]  user '%s' is signing in ...", login)
```

I am quite sure that you do not have the most recent version or that you may even not request on the right server ... could you check that please ?

**Comment by jbaptl** [comment]
Well done!
I found an `auth.py.sav` file on `/var/lib/shinken/modules/webui2/submodules` directory.
My fault! it was certainly a test i have done before opening the issue. I'm sorry for that.

Just in case, I have reset webui2 module with the following operations :

```
rm -rf /var/lib/shinken/modules/webui2
rm -rf /etc/shinken/modules/webui2.cfg
rm -rf /var/lib/shinken/inventory/webui2
su - shinken
shinken install webui2
service shinken restart
```

It doesn t correct the error, i still have the same.
But logs, now looks like this :

```
[1474529233] INFO: [broker-master] [WebUI]  user '' is signing in ...
[1474529233] INFO: [broker-master] [WebUI] Checking authentication for user: 
[1474529233] INFO: [broker-master] [WebUI] Requesting authentication for user: 
[1474529233] INFO: [broker-master] [WebUI] Authenticating user ''
[1474529233] INFO: [broker-master] [WebUI] Internal htpasswd authentication
[1474529233] INFO: [broker-master] [WebUI] Internal alignak backend authentication
[1474529233] INFO: [broker-master] [WebUI] Internal contact authentication
[1474529233] INFO: [broker-master] [WebUI-auth-cfg-password] Authenticating user ''
```

I can re-try a full basic installation, but i'm quite sure i will have the same error.

**Comment by mohierf** [comment]
Yesterday, for my tests, I made a fresh install for Shinken and WebUI2 and I did not got your error ... it looks like if your POST variables were not even seen by the WebUI Bottle server. 

For me it has nothing to do with the WebUI itself but with your configuration. Di you made the test locally with http://127.0.0.1:7767 ?

**Comment by jbaptl** [comment]
Detail of my full resettlement:

```
rm -rf /var/lib/shinken
rm -rf /etc/shinken
mv /var/log/shinken /var/log/savshinkenlog
rm -rf /var/run/shinken
rm -rf /usr/local/lib/python2.7/site-packages/Shinken-2.4.3-py2.7.egg
unzip shinken-2.4.3.zip
cd shinken-master
./clean.sh
python2.7 setup.py install
su - shinken
shinken --init
shinken --proxy=<my_proxy> install webui2
```

Add `webui2` module on `broker-master`

```
grep modules /etc/shinken/brokers/broker-master.cfg
    # Interesting modules that can be used:
    modules     webui2
```

I need to modify `/etc/init.d/shinken` file to take `python2.7` in charge :

```
vi /etc/init.d/shinken
<PYTHON=$LASTFOUND
>PYTHON="/usr/local/bin/python2.7"
```

I still have the problem. Do i made something wrong?

**Comment by jbaptl** [comment]
Yes! I have the same problem locally.

**Comment by mohierf** [comment]
Hum... except for a difference between the Python installed packages, I do not see. 
On your configuration:

> python2.7
> bottle-0.12.8
> pymongo-3.3.0
> requests-2.11.1
> arrow-0.8.0
> passlib-1.6.5
> CherryPy-8.1.0
> pycurl-7.43.0
> pbr-1.10.0

On my configuration:

```
arrow==0.7.0
bottle==0.12.9
CherryPy==5.4.0
passlib==1.6.5
pbr==1.10.0
pycrypto==2.6.1
pycurl==7.19.0
pymongo==3.2.2
requests==2.10.0
setproctitle==1.1.10
Shinken==2.4.3
```

Perharps something to deal with a too recent CherryPy ?

**Comment by jbaptl** [comment]
Hello,
I made a new install of `python2.7`. I had an error to install the readline library.
So I have had to install the package `readline-devel` and `ncurses-devel`. After that, the installation of `python2.7` succeded and the installation of Shinken and webui2 module succeded too.

Now, shinken webui is working with some error, but it will be easy to debug.
I close the issue
I really thank you for your help.

Regards
